src/compiler/checker.ts

```diff
@@ -22619,34 +22621,136 @@ namespace ts {
+                     *       x;  // B
+                     *       if ('B' === x.type) {  // second trigger
+                     *           return x;  // B
+                     *       }
+                     *       x;  // In the second time narrowing, the computed type would only have one(not Union), but should still be triggered to narrow type to never.
+                     *   }
+                     */
+                    // use declaredType is a quick way to resolve issue above -- when computed type is not union, the declared type is not chagned.
+                    // But it also bring some other issue, like #39114 and #39110. As long as the declared type is not proper, issues come.
+
+                    // and only declared union is too loose to judge it could be narrowed.
+                    /**
+                     * Condition 2: Index type with undefined/null
+                     *   function ffff1() {
+                     *       let a: Array<number>|undefined;
+                     *       a = [1];
+                     *       if(a){
+                     *           if (a[2] !== undefined) {
+                     *               a[1];
+                     *           } else {
+                     *               a[2];  // declared type is union, but here should not be narrowed to never.
+                     *           }
+                     *       }
+                     *   }
+                     */
+                    if (!(declaredType.flags & TypeFlags.Union)) {
+                        return false;
+                    }
+                    if ((<UnionType>declaredType).types.filter(t => !(t.flags & TypeFlags.Primitive)).length < 2) {
+                        return false;    // to pass condition 2.
+                    }
+                    // However, I still use declared type, and put it here. The main aim is for future improving.
+                    // We could deal with 'this' type and use ugly code to deal with Condition 2, then we could avoid use declared type at least for now
+                    // passing all tests and issue #39110/#39114, but does it deserve? Is the code clear enough?
+
+                    // if(isThisTypeParameter(computedType)){
+                    //     return false;
+                    // }
+                    // if(isElementAccessExpression(expr)){
+                    //     // the only purpose of these code is to omit above condition 2.
+                    //     const type = getReducedApparentType(computedType);
+                    //     if (type.flags & TypeFlags.Object) {
+                    //         const resolved = resolveStructuredTypeMembers(<ObjectType>type);
+                    //         if(resolved.stringIndexInfo || resolved.numberIndexInfo){
+                    //             return false;
+                    //         }
+                    //     }
+                    // }
+                    return true;

```

Comment:
Again, these are separate issues you’re discussing. Let’s aim to start cleaning this PR up and leave the discussion on GitHub instead of in comments—these 60 lines should be ~3 lines, and I still think “condition 2” should not be addressed in this PR.

Reply:
Sure, the comment could(and should) be removed, but let me explain a little about why I always try to stress `condition2`: to pass the test.

Although there is no such a test case directly, but one step in the whole test is bootstrap(do I use the term correctly? I mean to check typescript itself with the new compiler). 

And Typescript does have the similar code for a long time. But it is interesting to find that it could pass test today(would double check), so all things could be ~3lines, suprising for me.

Reply:
I agree that line looks weird, and I was hoping we could remove it too 😆 I still don’t understand that filter, so I’ll have to take a closer look at what it means. Thanks for pointing out the failure in nodeServer.ts. That will be useful to look at.